File: core/src/main/java/hunternif/voxarch/plan/legacy/Node.java
package hunternif.voxarch.plan.legacy;

import hunternif.voxarch.vector.Vec3;

/**
 * Contains some generic properties of a part of an {@link ArchPlan}.
 * @author Hunternif
 */
public class Node {
	private Room parent;
	
	/** Is interpreted differently for different kinds of nodes. */
	private final Vec3 origin;
	
	private final double rotationY;
	
	/** The type specifies a purpose for the node. It can be used by a generator
	 * to assign a particular architectural style or materials to it. */
	private String type = null;
	
	/** For use with incremental building. This means that this node will not be re-built. */
	private boolean isBuilt = false;
	
	/** The origin vector will be copied. */
	public Node(Room parent, Vec3 origin, double rotationY) {
		if (parent == this) throw new IllegalArgumentException("Node can't be a parent of iteslf");
		this.parent = parent;
		this.origin = new Vec3(origin);
		this.rotationY = rotationY;
	}
	/** The origin vector will be copied. */
	public Node(Vec3 origin) {
		this(null, origin, 0);
	}
	
	
	protected void setParent(Room parent) {
		this.parent = parent;
	}
	public Room getParent() {
		return parent;
	}
	
	public Vec3 getOrigin() {
		return origin;
	}
	
	public double getRotationY() {
		return rotationY;
	}
	
	public void setType(String type) {
		this.type = type;
	}
	public String getType() {
		return type;
	}

	public void setBuilt(boolean isBuilt) {
		this.isBuilt = isBuilt;
	}
	public boolean isBuilt() {
		return isBuilt;
	}
}

File: core/src/main/kotlin/hunternif/voxarch/plan/Node.kt
package hunternif.voxarch.plan

import hunternif.voxarch.builder.Builder
import hunternif.voxarch.util.INested
import hunternif.voxarch.vector.Vec3

/**
 * **Coordinate axes** are like in Minecraft, i.e. Y is up.
 *  * X: width - the longer side, defines direction.
 *  * Y: height (up)
 *  * Z: depth
 *
 *  **Rotations** are counterclockwise, 0 degrees is East. (Minecraft rotations
 *  are clockwise.) This was done for consistency with mainstream vector math.
 *
 * ```
 * Top-down view of the reference frame:
 * Y
 *  +---->X (East)
 *  |
 *  V
 *  Z
 * ```
 *
 * At the planning stage, coordinates are floating-point 3d vectors.
 * But ultimately all structures will be built into a voxel world where
 * coordinates are _integer_ 3d vectors. Bridging this gap can be confusing.
 * I use 2 conventions for measuring the voxel space: **"centric"** and
 * **"natural"**.
 *
 *
 * ## 1. "Centric" position and distance
 * Think of voxel blocks as points in 3d space, where every integer point
 * corresponds to a center of a block. This convention defines how to position
 * blocks in space and a how to measure distance between them.
 * It should be used by default, unless otherwise specified.
 *
 * * Distance between adjacent blocks is 1.
 * * A room of "centric" size 0 occupies 1 block.
 * * A wall of "centric" width 2 will occupy 3 points, i.e. 3 blocks.
 * * In a room of "centric" size 2, its cross-section will look like this:
 * 1 block of wall, 1 block of empty space, 1 block of wall.
 *
 *
 * ## 2. "Natural" distance
 * This is how a normal sane person would measure buildings in Minecraft:
 * "if it uses 2 blocks to build, its size is 2". This convention should only
 * used to measure distances, not position! It's useful at the building stage,
 * when describing the size of details like crenellations on a castle wall.
 *
 * * The room of "centric" size 2 (described above) will have a "natural" size 3.
 *
 * @param origin coordinates of nodes inside this Node are counted
 *               from this origin.
 */
open class Node(
    origin: Vec3
) : INested<Node> {
    var origin: Vec3 = origin.clone()
        set(value) { field.set(value) }  // keep the same instance

    /** Internal offset of the low-XZ corner.
     * Children are placed relative to parent's [origin], but parent's [start]
     * suggests where children should be placed.
     * By default, it's set so that origin is at the low-XYZ corner. */
    open var start: Vec3 = Vec3(0, 0, 0)
        set(value) { field.set(value) }  // keep the same instance

    final override var parent: Node? = null

    /** Rotation around Y axis CCW in degrees */
    open var rotationY: Double = 0.0

    /** If set, this builder class will have higher priority than BuilderConfig */
    var builder: Builder<*>? = null

    /** If true, the node will not be generated, only used as boundary for calculations. */
    var transparent: Boolean = false

    /** Children will only place voxels inside this mask. */
    var clipMask: ClipMask = ClipMask.OFF

    override val children: MutableList<Node> = mutableListOf()

    /**
     * Vector (width, height, depth) in "centric" coordinates,
     * not accounting for rotation.
     */
    open var size: Vec3 = Vec3(0, 0, 0)
        set(value) { field.set(value) } // keep the same instance
    /** "centric" width (X axis) */
    var width: Double
        get() = size.x
        set(value) { size.x = value }
    /** "centric" height (Y axis) */
    var height: Double
        get() = size.y
        set(value) { size.y = value }
    /** "centric" depth (Z axis) */
    var depth: Double
        get() = size.z
        set(value) { size.z = value }

    /**
     * Tags can be Style class names, or building part names etc.
     * They will be used to find a Builder with a particular architectural style,
     * or building materials.
     * Does not automatically apply to children; if you need inheritance, make sure
     * to add the tags to children manually.
     */
    val tags = LinkedHashSet<String>()

    fun addChild(child: Node, position: Vec3) {
        addChild(child)
        child.origin = position
    }

    /**
     * Returns boundaries defined by node size and its walls, if it has any.
     * Relative to local origin, not accounting for rotation.
     */
    open fun getGroundBoundaries(): List<GroundBoundary> {
        val corner = start
        return listOf(
            corner,
            corner.addZ(size.z),
            corner.add(size.x, 0.0, size.z),
            corner.addX(size.x),
            corner,
        ).zipWithNext()
    }

    constructor() : this(Vec3.ZERO)

    override fun toString(): String =
        "${this::class.java.simpleName} [${tags.joinToString(", ")}]"
}

typealias GroundBoundary = Pair<Vec3, Vec3>